### PR TITLE
[MM-44185] Add conditions for subscription validation

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

This PR adds more checks to the `validateSubscription` function, as well as adding more test cases.

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-44185

### CODE REVIEW

**Comment by mickmister** [comment]
Not sure why CI is failing:

```
cd webapp && /usr/local/bin/npm install
make: *** [Makefile:68: webapp/.npminstall] Terminated

Too long with no output (exceeded 10m0s): context deadline exceeded
```

**Comment by mickmister** [comment]
Fixed CI with most recent commit

**Review by jupenur — APPROVED** [review]

**Review by levb — APPROVED** [review]

**Comment by mickmister** [comment]
@DHaussermann Gentle ping on review for this

**Comment by DHaussermann** [comment]
@mickmister - Not sure if I'm seeing a deployment issue or if something else is happening here. I don't see a new filter as expected for `Comment Visibility` so I am unable to do any functional testing for this.

I have the matching commit deployed. Any thoughts here? Maybe I've misunderstood something here?

**Comment by mickmister** [comment]
@DHaussermann and I have discussed offline the scope of this PR, which does not include changes to comment visibility

**Comment by jupenur** [comment]
@mickmister following up on our brief chat yesterday, what's missing to get this merged? Just the QA review? cc @DHaussermann

**Comment by mickmister** [comment]
@jupenur Yes QA review is the next step here

**Comment by DHaussermann** [comment]
@mickmister I did a round of testing and this seems to be working as expected. 

Is there a reason this is not just a bool radio button?
image
Also - Can I try and tweak the wording on the option for clarity?

**Comment by billcrispin** [comment]
@DHaussermann what would you like me to check for here?

**Comment by asatkinson** [comment]
@DHaussermann what should I be checking for here?
- UI, default values etc.?
- my one comment would be does this need to be turned on by default, or just in high-security shops.  I can definitely see why there are cases not enabling this would be useful.  If this is just turned on will it break any existing subscriptions?

**Comment by DHaussermann** [comment]
I have discussed this with @mickmister. 
1. For now, a compromise would be to add the following text below the "approximate JQL" line: 
"**Note** that since you have not selected a security level filter, the subscription will only allow issues that have no security level assigned."
The text would then disappear once a security level is added. 

2. I think the label of the enforcement flag would be more clear if we set the text to: 
"Subscriptions will only include issues that have a security level assigned if the appropriate security level has been included as a filter" 

3. A correction from above 
> "With this design Security Level is no longer optional. The UI should reflect this when ...."

This is not true. Technically they are still optional. Without a filter only issues with no security level assigned are included. This is a valid use case.

cc @asatkinson

**Comment by mickmister** [comment]
@DHaussermann I've updated the PR with the changes mentioned above. Note that the sentence wraps around with one word on the second line. This looks a little asymmetrical/odd to me:

image

File: server/subscribe.go [hunk 1]
```diff
@@ -16,6 +16,8 @@ import (
 	jira "github.com/andygrunwald/go-jira"
 	"github.com/pkg/errors"
 
+	"github.com/trivago/tgo/tcontainer"
```
Comment (javaguirre) [inline server/subscribe.go:19]: This package doesn't have a big community, my worry is that It can become deprecated soon-ish in the future. Is the functionality it gives us very valuable? WDYT?

File: server/subscribe.go [hunk 2]
```diff
@@ -141,37 +143,53 @@ func (p *Plugin) matchesSubsciptionFilters(wh *webhook, filters SubscriptionFilt
...
 		containsAny := value.ContainsAny(field.Values.Elems()...)
 		containsAll := value.ContainsAll(field.Values.Elems()...)
 
-		if (field.Inclusion == FilterIncludeAny && !containsAny) ||
-			(field.Inclusion == FilterIncludeAll && !containsAll) ||
-			(field.Inclusion == FilterExcludeAny && containsAny) ||
-			(field.Inclusion == FilterEmpty && value.Len() > 0) {
-			validFilter = false
-			break
+		if (inclusion == FilterIncludeAny && !containsAny) ||
```
Comment (javaguirre) [inline server/subscribe.go:177]: This condition could be a method and then very easy to unittest to avoid breaking changes in the future and peace of mind. 👼

File: server/subscribe.go [hunk 3]
```diff
@@ -298,6 +316,37 @@ func (p *Plugin) validateSubscription(instanceID types.ID, subscription *Channel
 		return errors.New("please provide a project identifier")
 	}
 
+	projectKey := subscription.Filters.Projects.Elems()[0]
+
+	var securityLevels StringSet
+	useEmptySecurityLevel := p.getConfig().SecurityLevelEmptyForJiraSubscriptions
+	for _, field := range subscription.Filters.Fields {
+		if field.Key != securityLevelField {
```
Comment (javaguirre) [inline server/subscribe.go:324]: If the logic inside of the loop `for` is a method, it's very easy to unittest and branch coverage is easy to get.

**Review by javaguirre — CHANGES_REQUESTED** [review]
Let me know what you think 👍

File: server/subscribe.go (hunk 1 above)
Comment (mickmister) [reply server/subscribe.go:19]: I think we actually have to use this because it's the underlying data structure being used by https://github.com/andygrunwald/go-jira for arbitrary field types. I'm not sure if there's a way around this.

Comment (mickmister) [reply server/subscribe.go:19]: @javaguirre I think we actually have to use this because it's the underlying data structure being used by https://github.com/andygrunwald/go-jira for arbitrary field types. I'm not sure if there's a way around this.

**Comment by DHaussermann** [comment]
I have done a round of testing and this is looking good.
- Tested that security level filter is no longer opt-in on upgrade
- Tested that the "kill switch" works as expected to continue subscription delivery
- Regression tested security level filter is working
- Labeling is system console has been improved for clarity
- Subscription creation screen now has user feed back that there is now security level filter set when this is the case

When I tried to repeat this testing on server (testing above was on cloud) I get an error saving the subscription

image

@mickmister couple questions...
- Is there any chance this change is agnostic to Jira Server Vs Jira Cloud? to me this seems like a risk.
- The error above seems unrelated to the PR but can you confirm if it's happening on our side or on the Jira side?

**Comment by DHaussermann** [comment]
I tested this again and it is working as expected in this branch provided @mickmister here https://github.com/mattermost/mattermost-plugin-jira/tree/validate-subscription_debug.

**Comment by DHaussermann** [comment]
@m1lt0n or @asatkinson we should discuss the impact of this change and if clients need to be warned about it. 
This PR really should have some type of Business Analysts approval. 

Testing this has caused me to wonder how common the case is where people use projects in Jira with no issue security scheme applied. In such a case, users won't quickly be able to go edit their subscription to get it working again once they upgrade the plugin. Some effort is required in Jira as well to associate the project with a security scheme.

In short...
Issue where a scheme is applied and the security level is set to `none` works fine and no changes are needed to the subscription. But our solution here kind of assumes all projects have issue security schemes applied.   
For any project where there is no issue security scheme in use there will be no way of delivering subscriptions unless the "kill switch"  to make it work the old way is turned on to not require security level.

**Comment by asatkinson** [comment]
@DHaussermann I've been looking through this and I don't think I fully understand why this is so problematic.  
It seems like they just want a switch so that if this filter is set then we restrict subscription notification to those posts below a certain security level - as defined in this JIRA [help doc](https://support.atlassian.com/jira-cloud-administration/docs/configure-issue-security-schemes/)

Can we do a quick call so you can walk me through this?  I don't have access to this UI.

**Comment by asatkinson** [comment]
If the issue is that it applies this and prevents all other subscriptions from working, then yes, this is a big problem and we should find another way to do this.

**Comment by mickmister** [comment]
@DHaussermann Gentle ping to give this PR another look. I've resolved issues in CI

**Comment by DHaussermann** [comment]
After discussing with @mickmister again the last concern I had is not applicable here. 

I have checked again and tested on both Cloud and Server. If the issue is part of a project where there is no security level to assign because no level is in use by the project, the subscription **will** deliver the event even though the subscription has no applicable security level to evaluate.

**Review by DHaussermann — APPROVED** [review]
Tested and passed
 
- On upgrade, subscription with security level continues to work
- On upgrade, you can add a security level to a subscription to make it start delivering again
- On Upgrade there is valid feedback that Exclude operator can no longer be used for security level when user open edit view to see why it's not working. 
image
- "Kill switch" works to revert to old logic
- Updating a subscription to add security level works 
- Targeting a project where there is no applicable security level works as expected without the filter as none is applicable

LGTM!

Thanks @mickmister for the effort on this.
